### ISSUE TITLE
Unable to Unpack TypedDict-bound TypeVar

### ISSUE THREAD

**Issue by JP-Ellis** [python/mypy#16525]
**Bug Report**

Mypy fails to recognize type variables which are bound to a `TypedDict` when subsequently used with `Unpack`.

**To Reproduce**

```python
from typing import TypedDict, TypeVar, Unpack

Data = TypeVar("Data", bound=TypedDict)
# Data = TypeVar("Data", bound=type[TypedDict])  # also fails


def foo(**kwargs: Unpack[Data]):
    pass
```

[mypy-play link](https://mypy-play.net/?mypy=master&python=3.12&gist=6dbd6ad37793c01f5bda5c0cd3714e73)

**Expected Behavior**

Ideally, Mypy should recognise that `Data` is expected to be an instance of a `TypedDict` and therefore when `Unpack` is used, no error message is reported.

**Actual Behavior**

Mypy is unable to recognise that `Data` is an instance of a `TypedDict`.

The reported errors are:

```log
main.py:3: error: Variable "typing.TypedDict" is not valid as a type  [valid-type]
main.py:3: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
main.py:7: error: Unpack item in ** argument must be a TypedDict  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

I also note that trying to address the first error message with `type[typing.TypedDict]` does not work either.

**Your Environment**

Discovered in:

- Mypy version used: 1.7.0
- Mypy command-line flags: _none_
- Mypy configuration options from `mypy.ini` (and other config files): _none_
- Python version used: 3.11.6

Based on the Mypy playground, the error also exists in master against Python 3.12.

**Comment by erictraut** [comment]
I think mypy's behavior is correct here. There are a couple of problems with your code sample, both of which mypy identifies:

1. [PEP 692](https://peps.python.org/pep-0692/) is clear that only a `TypedDict` type can be used in the `Unpack`. A TypeVar bound to a `TypedDict` is not allowed. It's important for an unpacked `TypedDict` to be concrete when describing a `**kwargs` type.

2. Creating a TypeVar that is bound to `TypedDict` doesn't make sense. `TypedDict` is not a base class and doesn't participate in class hierarchies. It's a special form that is used to define a structural type. It's similar to `Protocol`, which similarly doesn't make sense to use as an upper bound in a `TypeVar`.

**Comment by JP-Ellis** [comment]
Thanks for being super quick to reply!

Regarding point (2) you raise, I think there is a circumstance where it might make sense. My specific use case came from defining an abstract base class that has a generic `TypedDict` associated to it. I.e.,

```python
class AbstractFoo[Data: TypedDict](ABC):
    @abstractmethod
    def some_method(self, **kwargs: Unpack[Data]):
        pass

class ConcreteTypedDict(TypedDict):
    a: int
    b: str
    c: bool

class MyFoo(AbstractFoo[ConcreteTypedDict]):
    def some_method(self, **kwargs: Unpack[ConcreteTypedDict]):
        assert kwargs["c"]
```

The abstract class is never instantiated, but once the concrete subclass is instantiated, there is a definite `TypedDict` associated with the class.

I removed the classes from the example in order to make the example minimal; however, given this intent, do you know whether this is still a bug? Or working as intended?

**Comment by erictraut** [comment]
> I think there is a circumstance where it might make sense

I don't think so. As I said, `TypedDict` doesn't participate in type hierarchies. It's a special form that is valid only within a class definition statement. It indicates that you're defining a structural type, similar to `Protocol`. From a type theory perspective, it doesn't make sense to allow `TypedDict` as an upper bound for a TypeVar. Nor does it make sense for it to appear in a type annotation.

> do you know whether this is still a bug?

I don't think it's a bug. I think mypy is working in accordance with type theory and with [PEP 589](https://peps.python.org/pep-0589/) (which introduced `TypedDict`) and [PEP 692](https://peps.python.org/pep-0692/) (which introduced using TypedDict for more precise kwargs typing).
